File: android/app/src/main/java/com/alexivanov/snapsell/data/repository/InventoryRepository.kt
package com.alexivanov.snapsell.data.repository

import com.alexivanov.snapsell.core.Clock
import com.alexivanov.snapsell.core.Ids
import com.alexivanov.snapsell.data.local.ItemEntity
import com.alexivanov.snapsell.data.local.ListingEntity
import com.alexivanov.snapsell.data.local.ListingWithItems
import com.alexivanov.snapsell.data.local.SnapsellDatabase
import com.alexivanov.snapsell.data.remote.dto.PriceQuote
import com.alexivanov.snapsell.domain.ListingKind
import com.alexivanov.snapsell.domain.ListingStatus
import com.alexivanov.snapsell.domain.SuggestedPrice
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first

/** On-device inventory (Room). Every write stamps updatedAt from [clock]. */
class InventoryRepository(
    private val db: SnapsellDatabase,
    private val clock: Clock,
) {
    private val items get() = db.itemDao()
    private val listings get() = db.listingDao()

    fun observeItems(): Flow<List<ItemEntity>> = items.observeAll()
    fun observeItem(id: String): Flow<ItemEntity?> = items.observe(id)
    suspend fun getItem(id: String): ItemEntity? = items.get(id)
    suspend fun getItems(ids: List<String>): List<ItemEntity> {
        // IN (...) does not preserve order; callers care about it.
        val byId = items.getByIds(ids).associateBy { it.id }
        return ids.mapNotNull { byId[it] }
    }

    suspend fun saveItem(item: ItemEntity) = items.upsert(item.copy(updatedAt = clock.nowMillis()))
    suspend fun saveItems(list: List<ItemEntity>) = items.upsertAll(list.map { it.copy(updatedAt = clock.nowMillis()) })
    suspend fun deleteItem(id: String) = items.delete(id)

    /** Stores a fresh quote. A final price the user typed is never touched. */
    suspend fun saveQuote(itemId: String, quote: PriceQuote) {
        val existing = items.get(itemId) ?: return
        items.upsert(existing.copy(quote = quote, updatedAt = clock.nowMillis()))
    }

    /** null clears the override so the suggestion shows again. */
    suspend fun setFinalPrice(itemId: String, price: Double?) {
        val existing = items.get(itemId) ?: return
        items.upsert(existing.copy(finalPrice = price, updatedAt = clock.nowMillis()))
    }

    /**
     * Settings changed the local-sale factor: every stored quote gets a new
     * suggested_price from its own asking_median, locally. Final prices are
     * user input and are left alone.
     */
    suspend fun recomputeSuggestedPrices(localSaleFactor: Double) {
        val all = items.observeAll().first()
        val now = clock.nowMillis()
        val updated = all.mapNotNull { item ->
            val quote = item.quote ?: return@mapNotNull null
            val next = quote.copy(
                suggestedPrice = SuggestedPrice.recompute(quote.askingMedian, localSaleFactor),
                localSaleFactor = localSaleFactor,
            )
            if (next == quote) null else item.copy(quote = next, updatedAt = now)
        }
        if (updated.isNotEmpty()) items.upsertAll(updated)
    }

    fun observeListings(): Flow<List<ListingWithItems>> = listings.observeAllWithItems()
    fun observeListing(id: String): Flow<ListingWithItems?> = listings.observeWithItems(id)
    suspend fun getListing(id: String): ListingWithItems? = listings.getWithItems(id)

    /** Creates a draft listing and returns its id. */
    suspend fun createListing(
        kind: ListingKind,
        itemIds: List<String>,
        title: String,
        description: String,
        price: Double,
        bundleDiscount: Double? = null,
    ): String {
        val now = clock.nowMillis()
        val id = Ids.newId()
        listings.insertWithItems(
            ListingEntity(
                id = id,
                kind = kind,
                title = title,
                description = description,
                price = price,
                bundleDiscount = bundleDiscount,
                status = ListingStatus.DRAFT,
                createdAt = now,
                updatedAt = now,
            ),
            itemIds,
        )
        return id
    }

    suspend fun updateListingStatus(id: String, status: ListingStatus) {
        val existing = listings.getWithItems(id)?.listing ?: return
        val now = clock.nowMillis()
        listings.updateStatus(
            id = id,
            status = status,
            listedAt = if (status == ListingStatus.LISTED) existing.listedAt ?: now else existing.listedAt,
            soldAt = if (status == ListingStatus.SOLD) existing.soldAt ?: now else existing.soldAt,
            updatedAt = now,
        )
    }
}

File: ios/SnapSell/Data/Repository/InventoryRepository.swift
import Foundation
import Observation
import SwiftData

/// On-device inventory (SwiftData). Every write stamps updatedAt from `clock`.
/// `items` and `listings` are observable snapshots, newest first, refreshed
/// after every write; views and view models read them and re-render.
@MainActor
@Observable
final class InventoryRepository {
    private(set) var items: [Item] = []
    private(set) var listings: [ListingWithItems] = []

    @ObservationIgnored private let context: ModelContext
    @ObservationIgnored private let clock: Clock

    init(container: ModelContainer, clock: Clock) {
        self.context = container.mainContext
        self.clock = clock
        reload()
    }

    // ---------------------------------------------------------------- reads

    func item(_ id: String) -> Item? { items.first { $0.id == id } }

    func listing(_ id: String) -> ListingWithItems? { listings.first { $0.id == id } }

    /// Items in the order of `ids`; unknown ids are skipped.
    func items(_ ids: [String]) -> [Item] {
        let byId = Dictionary(items.map { ($0.id, $0) }, uniquingKeysWith: { a, _ in a })
        return ids.compactMap { byId[$0] }
    }

    // ---------------------------------------------------------------- items

    func saveItem(_ item: Item) throws {
        var stamped = item
        stamped.updatedAt = clock.nowMillis()
        try upsert(stamped)
        try commit()
    }

    func saveItems(_ list: [Item]) throws {
        let now = clock.nowMillis()
        for var item in list {
            item.updatedAt = now
            try upsert(item)
        }
        try commit()
    }

    func deleteItem(_ id: String) throws {
        if let record = try fetchItem(id) { context.delete(record) }
        try commit()
    }

    /// Stores a fresh quote. A final price the user typed is never touched.
    func saveQuote(itemId: String, quote: PriceQuote) throws {
        guard var existing = item(itemId) else { return }
        existing.quote = quote
        try saveItem(existing)
    }

    /// nil clears the override so the suggestion shows again.
    func setFinalPrice(itemId: String, price: Double?) throws {
        guard var existing = item(itemId) else { return }
        existing.finalPrice = price
        try saveItem(existing)
    }

    /// Settings changed the local-sale factor: every stored quote gets a new
    /// suggested_price from its own asking_median, locally. Final prices are
    /// user input and are left alone.
    func recomputeSuggestedPrices(localSaleFactor: Double) throws {
        let now = clock.nowMillis()
        var changed = false
        for var item in items {
            guard var quote = item.quote else { continue }
            quote.suggestedPrice = try SuggestedPrice.recompute(askingMedian: quote.askingMedian, localSaleFactor: localSaleFactor)
            quote.localSaleFactor = localSaleFactor
            if quote != item.quote {
                item.quote = quote
                item.updatedAt = now
                try upsert(item)
                changed = true
            }
        }
        if changed { try commit() }
    }

    // ---------------------------------------------------------------- listings

    /// Creates a draft listing and returns its id.
    func createListing(kind: ListingKind, itemIds: [String], title: String, description: String, price: Double, bundleDiscount: Double? = nil) throws -> String {
        let now = clock.nowMillis()
        let id = Ids.newId()
        let listing = Listing(
            id: id, kind: kind, itemIds: itemIds, title: title, description: description, price: price,
            bundleDiscount: bundleDiscount, status: .draft, createdAt: now, updatedAt: now
        )
        context.insert(ListingRecord(listing))
        try commit()
        return id
    }

    func updateListingStatus(_ id: String, status: ListingStatus) throws {
        guard let record = try fetchListing(id) else { return }
        let now = clock.nowMillis()
        record.status = status.wire
        if status == .listed, record.listedAt == nil { record.listedAt = now }
        if status == .sold, record.soldAt == nil { record.soldAt = now }
        record.updatedAt = now
        try commit()
    }

    // ---------------------------------------------------------------- internals

    private func upsert(_ item: Item) throws {
        if let record = try fetchItem(item.id) {
            record.apply(item)
        } else {
            context.insert(ItemRecord(item))
        }
    }

    private func fetchItem(_ id: String) throws -> ItemRecord? {
        var d = FetchDescriptor<ItemRecord>(predicate: #Predicate { $0.id == id })
        d.fetchLimit = 1
        return try context.fetch(d).first
    }

    private func fetchListing(_ id: String) throws -> ListingRecord? {
        var d = FetchDescriptor<ListingRecord>(predicate: #Predicate { $0.id == id })
        d.fetchLimit = 1
        return try context.fetch(d).first
    }

    private func commit() throws {
        try context.save()
        reload()
    }

    private func reload() {
        let itemRecords = (try? context.fetch(FetchDescriptor<ItemRecord>(sortBy: [SortDescriptor(\.createdAt, order: .reverse)]))) ?? []
        let all = itemRecords.map(\.item)
        let byId = Dictionary(all.map { ($0.id, $0) }, uniquingKeysWith: { a, _ in a })
        let listingRecords = (try? context.fetch(FetchDescriptor<ListingRecord>(sortBy: [SortDescriptor(\.createdAt, order: .reverse)]))) ?? []
        items = all
        listings = listingRecords.map { r in
            let l = r.listing
            return ListingWithItems(listing: l, items: l.itemIds.compactMap { byId[$0] })
        }
    }
}
